sonic_platform_base/sonic_xcvr/fields/xcvr_field.py

```diff
@@ -0,0 +1,148 @@
+
+
+class XcvrField(object):
+    def __init__(self, name, offset, ro):
+        self.name = name
+        self.offset = offset
+        self.ro = ro
+
+    def get_fields(self):
+        fields = {}
+        if hasattr(self, "fields"):
+            for field in self.fields:
+                fields[field.name] = field
+                fields.update(field.get_fields())
+        return fields
+
+    def get_offset(self):
+        return self.offset
+
+    def get_size(self):
+        raise NotImplementedError
+
+    def read_before_write(self):
+        raise NotImplementedError
+
+    def decode(self, raw_data):
+        raise NotImplementedError
+
+    def encode(self, val, raw_state=None):
+        raise NotImplementedError
+
+
+class RegBitField(XcvrField):
+    def __init__(self, name, bitpos, parent=None, ro=True, offset=None):
+        super(RegBitField, self).__init__(name, offset, ro)
+        self.bitpos = bitpos
+        self.parent = parent
+
+    def get_size(self):
+        return 1
+
+    def read_before_write(self):
+        return True
+
+    def decode(self, raw_data):
+        return (raw_data[0] >> self.bitpos) & 1
+
+    def encode(self, val, raw_state=None):
+        assert not self.ro and raw_state is not None
+        curr_state = raw_state[0]
+        if val:
+            curr_state |= (1 << self.bitpos)
+        else:
+            curr_state &= ~(1 << self.bitpos)
+        return bytearray([curr_state])
+
+
+class RegField(XcvrField):
+    def __init__(self, name, offset, *fields, **kwargs):
+        super(RegField, self).__init__(name, offset, kwargs.get("ro", True))
+        self.fields = fields
+        self.size = kwargs.get("size", 1)
+        self._updateBitOffsets()
+
+    def _updateBitOffsets(self):
+        for field in self.fields:
+            field.offset = self.offset + field.bitpos / 8
+
+    def get_size(self):
+        return self.size
+
+    def read_before_write(self):
+        return False
+
+class NumberRegField(RegField):
+    def __init__(self, name, offset, *fields, **kwargs):
+        super(NumberRegField, self).__init__( name, offset, *fields, **kwargs)
+        self.scale = kwargs.get("scale")
+        self.format = kwargs.get("format", ">B")
+
+    def decode(self, raw_data):
+        decoded = struct.unpack(self.format, raw_data)[0]
+        if self.scale is not None:
+            assert self.scale != 0
+            return decoded / self.scale
+        return decoded 
+
+    def encode(self, val, raw_state=None):
+        assert not self.ro
+        if self.scale is not None:
+            return bytearray(struct.pack(self.format, val * self.scale))
+        return bytearray(struct.pack(self.format, val))
+
+class StringRegField(RegField):
+    def __init__(self, name, offset, *fields, **kwargs):
+        super(StringRegField, self).__init__(name, offset, *fields, **kwargs)
+        self.encoding = kwargs.get("encoding", "ascii")
+        self.format = kwargs.get("format", ">%ss" % self.size)
```

Comment:
`self.size` is an integer. Should the default format string not be `">%ds"`?

Reply:
Indeed, good catch.